File: lib/multi_modal_transformer.dart
// file: multimodal_transformer.dart

import 'gpu_tensor.dart';
import 'audio_transformer.dart';
import 'nn.dart';
import 'video_transformer.dart';

class MultimodalTransformer extends Module {
  final AudioTransformer audioModel;
  final VideoTransformer videoModel;
  final Layer fusionHead;
  final int numClasses;

  MultimodalTransformer({
    required this.audioModel,
    required this.videoModel,
    required this.numClasses,
  }) : fusionHead = Layer(
         audioModel.embedSize + videoModel.embedSize, // 64 + 128 = 192
         numClasses,
         useGelu: false,
       );

  /// Forward pass performing 'Intermediate Fusion' on GPU
  Tensor forward(Tensor audioIn, Tensor videoIn, List<Tensor> tracker) {
    // 1. Extract Audio Features
    final audioEmbed = audioModel.featureProjection.forward(audioIn, tracker);
    final aPos = audioModel.posEmbeddings.slice(0, audioIn.shape[0]);
    final aCombined = audioEmbed + aPos;
    tracker.add(aCombined);
    final aEncoded = audioModel.transformerEncoder.forwardEmbeddings(
      aCombined,
      tracker,
    );
    final aPooled = aEncoded.mean(); // [1, 64]
    tracker.add(aPooled);

    // 2. Extract Video Features
    Tensor vProjected = videoModel.frameProjection != null
        ? videoModel.frameProjection!.forward(videoIn, tracker)
        : videoIn;
    final vPos = videoModel.posEmbeddings.slice(0, videoIn.shape[0]);
    final vCombined = vProjected + vPos;
    tracker.add(vCombined);
    final vEncoded = videoModel.transformerEncoder.forwardEmbeddings(
      vCombined,
      tracker,
    );
    final vPooled = vEncoded.mean(); // [1, 128]
    tracker.add(vPooled);

    // 3. Fusion via Concatenation [1, 192]
    final fused = Tensor.concat([aPooled, vPooled]);
    tracker.add(fused);

    // 4. Final Classification
    return fusionHead.forward(fused, tracker);
  }

  @override
  List<Tensor> parameters() => [
    ...audioModel.parameters(),
    ...videoModel.parameters(),
    ...fusionHead.parameters(),
  ];
}

File: python/dart_cuda/core/transformers/modalities/multi_modal_transformer.py
"""Multi-modal transformer (audio + video) — mirrors
`modalities/multi_modal_transformer.dart`."""

from __future__ import annotations

from typing import List

from ...layers.nn import Layer, Module
from ...tensor.gpu_tensor import Tensor
from .audio_transformer import AudioTransformer
from .video_transformer import VideoTransformer


class MultimodalTransformer(Module):
    def __init__(
        self,
        *,
        audio_model: AudioTransformer,
        video_model: VideoTransformer,
        numClasses: int,
    ):
        self.audio_model = audio_model
        self.video_model = video_model
        self.num_classes = numClasses
        self.fusion_head = Layer(
            audio_model.embed_size + video_model.embed_size,
            numClasses,
            useGelu=False,
        )

    def forward(self, audio_in: Tensor, video_in: Tensor, tracker: List[Tensor]) -> Tensor:
        a_embed = self.audio_model.feature_projection.forward(audio_in, tracker)
        a_pos = self.audio_model.pos_embeddings.slice(0, audio_in.shape[0])
        a_combined = a_embed + a_pos
        tracker.append(a_combined)
        a_encoded = self.audio_model.transformer_encoder.forward_embeddings(a_combined, tracker)
        a_pooled = a_encoded.mean()
        tracker.append(a_pooled)

        if self.video_model.frame_projection is not None:
            v_proj = self.video_model.frame_projection.forward(video_in, tracker)
        else:
            v_proj = video_in
        v_pos = self.video_model.pos_embeddings.slice(0, video_in.shape[0])
        v_combined = v_proj + v_pos
        tracker.append(v_combined)
        v_encoded = self.video_model.transformer_encoder.forward_embeddings(v_combined, tracker)
        v_pooled = v_encoded.mean()
        tracker.append(v_pooled)

        fused = Tensor.concat([a_pooled, v_pooled])
        tracker.append(fused)
        return self.fusion_head.forward(fused, tracker)

    def parameters(self) -> List[Tensor]:
        return [
            *self.audio_model.parameters(),
            *self.video_model.parameters(),
            *self.fusion_head.parameters(),
        ]
